### PR TITLE
support lazy construction of WebSocketCommonProtocol

### PR DESCRIPTION
**Pull request by graingert**

### CODE REVIEW

File: src/websockets/legacy/protocol.py [hunk 1]
```diff
@@ -216,6 +207,30 @@ def __init__(
...
         self.close_connection_task: asyncio.Task[None]
+            
+    @property
+    def loop(self) -> asyncio.AbstractEventLoop:
+        return asyncio.get_running_loop()
+    
+    @loop.setter
+    def loop(self, value: asyncio.AbstractEventLoop) -> None:
+        if value is not asyncio.get_running_loop():
+            raise AttributeError("can't set attribute)
```
Comment (graingert) [inline src/websockets/legacy/protocol.py:218]:
```suggestion
            raise AttributeError("can't set attribute")
```

**Comment by graingert** [comment]
uvicorn constructs WebSocketCommonProtocol objects before the loop starts running

**Comment by aaugustin** [comment]
I understand you have a problem and I'd like to help you resolve it.

However, your explanation of the problem is quite terse:

> uvicorn constructs WebSocketCommonProtocol objects before the loop starts running, defer retrieval of the loop until absolutely necessary (when connection_made is called)

Furthermore, the patch quite large and includes potentially backwards incompatible changes where consequences are hard to assess, like this one:

```diff
-        self._drain_lock = asyncio.Lock(**loop_if_py_lt_38(loop))
+        self._drain_lock = asyncio.Lock()
```

You're also removing some attributes that used to be set in the constructor. This will definitely cause trouble; similar changes caused #767 previously. Let's not make this mistake again.

It seems to me that you're getting burnt by the change in Python 3.10 that makes many asyncio APIs non functional unless an event loop is running. This change was quite painful for websockets as it's test suite relied heavily on doing setup, then running the event loop, then checking results.

Could we discuss further the problem you're facing, review possible solutions, and look for a design that creates fewer backwards-compatibility risks?

**Comment by aaugustin** [comment]
Is this related to https://github.com/encode/uvicorn/pull/1070?

**Comment by graingert** [comment]
@aaugustin yep that's the PR

**Comment by aaugustin** [comment]
I see that you continue pushing code changes so I'm afraid I didn't get the right message across.

Currently this PR is headed the wrong way. I cannot merge it because it contains too many changes that I disagree with and where we didn't seek consensus on the design, even after I asked you to stop coding and start discussions. My concerns include but aren't limited to:

- taking risks with backwards compatibility in the `legacy` module, including but not limited to:
    - changing attributes of protocols
    - changing sequences of events — it's basically impossible to tell if these changes can cause side effects in review
- moving things around without a clear reasoning and a clear plan
- `if TYPE_CHECKING`

**Comment by aaugustin** [comment]
Please file an issue explaining the problem you're trying to solve, so we can figure out a solution. Thank you.

**Pull request closed without merging**